### PR TITLE
File musical chairs

### PR DESCRIPTION
**Pull request by dwalton76**
For issue #375 

@dlech @ddemidov @WasabiFan @ensonic take a look, this is a first stab at getting everything split up and moved around.  TouchSensor, Motors, ColorSensor, Buttons, InfraredSensor, and LEDs are all working.

### CODE REVIEW

**Comment by dwalton76** [comment]
Ignoring the fonts directory, this is the current structure
```
robot@evb[ev3dev]# tree .
.
├── _button.py
├── button.py
├── control
│   ├── GyroBalancer.py
│   ├── __init__.py
│   ├── rc_tank.py
│   └── webserver.py
├── display.py
├── fonts
│   ├── charB08.pbm


│   ├── timR18.pil
│   ├── timR24.pbm
│   └── timR24.pil
├── helper.py
├── __init__.py
├── _led.py
├── led.py
├── motor.py
├── _platform
│   ├── brickpi.py
│   ├── ev3.py
│   ├── evb.py
│   ├── __init__.py
│   └── pistorms.py
├── port.py
├── sensor
│   ├── __init__.py
│   └── lego.py
├── sound.py
└── version.py

4 directories, 423 files
robot@evb[ev3dev]# 
```

**Comment by dwalton76** [comment]
Can we make travis have a `/sys/class/board-info/`? See CI failure...

**Comment by WasabiFan** [comment]
Why are there two files for buttons?

**Comment by dlech** [comment]
version could be put in the root `__init__.py`, so there is no need for a module for just that.

**Comment by dlech** [comment]
fonts probably shouldn't be in the source code tree.

**Comment by dlech** [comment]
Is there anything left in `helper.py` after moving stuff to `control`?

**Comment by dlech** [comment]
It seems like it would be cleaner to do some import magic in `_platform/__init__.py` rather than having `button`/`_button` and `led`/`_led`

**Comment by dwalton76** [comment]
- `version.py` is generated via `python3 setup.py install`
- `fonts` not sure what the scoop is there, I've never used them
- `helper.py` I'll go ahead and chop it, I just need to redo the MINDCUBER demo but at this point we need to redo all of the demos so no point holding off

**Comment by dwalton76** [comment]
yeah button and led need some work that was just what I came up with to get the importing working.  These are both things where a platform specific class must live in `_platform` and that class must inherit from some parent class.  So button for instance
- we know we want the user to `from ev3dev.button import Button`
- The Button class will live in `_platform` but needs `ButtonEVIO`
- So where to put `ButtonEVIO`? It can't be in `button.py` else you get into a circular import situation between it and the `_platform` file.

**Comment by dwalton76** [comment]
@ddemidov thoughts on `fonts`?

**Comment by dlech** [comment]
Why would we have platform-specific button classes? As far as I can tell, the only thing that should be different is the path to the device node.

**Comment by dwalton76** [comment]
There is no guarantee that all platforms will have the same buttons...it just happens that EV3 and EVB do.  And since pistorms doesn't have any it shouldn't really have a Button class.

**Comment by dwalton76** [comment]
Wait I think pistorms has a `back` button but that is it and brickpi doesn't have any buttons.

**Comment by dlech** [comment]
How will documentation work if the same class is not the same between platforms? It seems to me like it would be better to make them all the same and throw a `RuntimeError` if something is missing (and the documentation will say something like "raises RuntimeError on BrickPi" if a class or method is not supported on that platform).

P.S. PiStorms has a Go button. (And one can add buttons to BrickPi using extra gpios)

**Comment by dwalton76** [comment]
I'll give you it would certainly clean up the importing. What about LEDs though they differ a good deal from EV3 vs pistorms?

**Comment by ddemidov** [comment]
Out of interest, what does `time python3 -c 'import ev3dev.ev3'` (before the PR) and `time python3 -c 'import ev3dev.motor'` (after the PR) show?

**Comment by ddemidov** [comment]
> thoughts on fonts?

I don't see why we could not leave them to be as is, except they are only usable with python3/PIL.

**Comment by ddemidov** [comment]
> Can we make travis have a /sys/class/board-info/? See CI failure...

We could put a `board-info` file into [ddemidov/ev3dev-lang-fake-sys](https://github.com/ddemidov/ev3dev-lang-fake-sys) (btw, should we move that repo under ev3dev?). Then we need to make the [get_current_platform](https://github.com/rhempel/ev3dev-lang-python/blob/799920eef3d8f34528243a40ccbb9d234782e9f6/ev3dev/core.py#L65) function to respect [Device.DEVICE_ROOT_PATH](https://github.com/rhempel/ev3dev-lang-python/blob/799920eef3d8f34528243a40ccbb9d234782e9f6/ev3dev/core.py#L150) constant (or, better yet, move the constant somewhere else). This way the travis tests [will know](https://github.com/rhempel/ev3dev-lang-python/blob/799920eef3d8f34528243a40ccbb9d234782e9f6/tests/api_tests.py#L14) where to look for the file.

**Comment by ddemidov** [comment]
See #225 for some background on fonts.

**Comment by ddemidov** [comment]
@dlech, 

> fonts probably shouldn't be in the source code tree

Do you mean we should put them into a separate package?

**Comment by ddemidov** [comment]
Re platform specific buttons: platform file could provide both path for the event file and the list of supported buttons. Since python is a dynamic language, we could use that list (during import) to create properties corresponding to the buttons inside `Button` class.

**Comment by dwalton76** [comment]
## Before
```
robot@ev3dev[~]# time python3 -c 'import ev3dev.ev3'

real	0m14.938s
user	0m11.340s
sys	0m0.560s
robot@ev3dev[~]# 
```

## After
```
robot@ev3dev[~]# time python3 -c 'import ev3dev.motor' 

real	0m11.645s
user	0m9.090s
sys	0m0.510s
robot@ev3dev[~]# 
```

**Comment by dwalton76** [comment]
Latest commit removes `_button.py` with the assumption that all platforms have the same buttons.  I'll test later on a pistorms and have it raise some NoButton exception if the user tries to use buttons on a device that doesn't have them.  Do you guys like the way this looks more?  If so I'll do the same for LEDs.

**Comment by dwalton76** [comment]
> btw, should we move that repo under ev3dev?

yes

**Comment by dwalton76** [comment]
> We could put a board-info file into ddemidov/ev3dev-lang-fake-sys

cool, @ddemidov can you add that? I haven't done much with ev3dev-lang-fake-sys

**Comment by ddemidov** [comment]
What should be in there? Should `FAKE-SYS` work?

**Comment by dwalton76** [comment]
Works for me

**Comment by dwalton76** [comment]
Should we move 'PowerSupply' out of '__init__.py' and into a 'power.py'?

**Comment by WasabiFan** [comment]
> Should we move 'PowerSupply' out of 'init.py' and into a 'power.py'?

I'd say "yes", because it's pretty rare that someone would use the power supply class. In my mind it's just clutter for most use cases. (In fact, I think we should try to minimize the classes in _init.py_ overall)

**Comment by dlech** [comment]
> What about LEDs though they differ a good deal from EV3 vs pistorms?

How so? PiStorms just has one extra color. I wouldn't call that a big difference.

**Comment by dlech** [comment]
> @dlech,
> 
>     fonts probably shouldn't be in the source code tree
> 
> Do you mean we should put them into a separate package?

I'm not sure what is the usual way for distributing data files with python packages, but is just seems odd to have data files mixed in with source code files in the source repository. Maybe better to save this discussion for later though.

File: ev3dev/_platform/ev3.py [hunk 1]
```diff
@@ -39,6 +38,8 @@
 INPUT_3 = 'in3'
 INPUT_4 = 'in4'
 
+buttons_filename = '/dev/input/by-path/platform-gpio_keys-event'
```
Comment (dlech) [inline ev3dev/_platform/ev3.py]: All caps for constants?

**Comment by dwalton76** [comment]
I was thinking EV3 had more LEDs than pistorms but now that I look at http://www.ev3dev.org/docs/platform-comparison/ I see that it doesn't.  I'll clean up LEDs to follow the same model as buttons.

File: ev3dev/_platform/ev3.py (hunk 1 above)
Comment (dwalton76) [reply ev3dev/_platform/ev3.py]: k

File: ev3dev/button.py [hunk 2]
```diff
@@ -28,14 +28,283 @@
 if sys.version_info < (3,4):
     raise SystemError('Must be using Python 3.4 or higher')
 
+import array
+import time
+import evdev
```
Comment (dlech) [inline ev3dev/button.py]: `evdev` is probably a linux-specific module like `fcntl`

**Comment by dlech** [comment]
> Since python is a dynamic language, we could use that list (during import) to create properties corresponding to the buttons inside Button class.

(Warning, getting a bit off topic here)

:+1: for taking advantage of dynamic nature of python to adapt to platforms, but I think I would rather see a single function that works for all buttons rather than identical functions/properties for each button. For example, I have a USB keypad that I can use for BrickPi/PiStorms that has the essential 6 EV3 buttons plus a bunch more. If I wanted to take advantage of other buttons, say the 10 numeric digits, then I would have to write 10 functions, one to handle each button. Whereas, if the button key code is just passed as a parameter, then I just have to write one function with a bunch of `elif`s.

File: ev3dev/button.py (hunk 2 above)
Comment (dwalton76) [reply ev3dev/button.py]: I was meaning to ask about the try/except around fcntl...do we have folks running `api_test.py` on windows?

Comment (dlech) [reply ev3dev/button.py]: Probably not, but any editor with code completion will load the modules.

**Comment by dwalton76** [comment]
Working on LEDs now...

**Comment by dwalton76** [comment]
LEDs are cleaned up, I tested them on EV3 and tried testing on pistorms but ran into https://github.com/ev3dev/ev3dev/issues/978

I think this is far enough along to merge and then we can tackle anything else via separate issues.  Will merge tomorrow unless folks want me to hold off.

**Comment by dwalton76** [comment]
Also resolves #379

**Comment by ddemidov** [comment]
>> We could put a board-info file into ddemidov/ev3dev-lang-fake-sys

> cool, @ddemidov can you add that? I haven't done much with ev3dev-lang-fake-sys

Done in https://github.com/ddemidov/ev3dev-lang-fake-sys/commit/61d8dedecd55614331aa0e6c1d6f180c4a4d949c

**Comment by WasabiFan** [comment]
OK, I understand what's going on. Inversion is now only done [here](https://github.com/gregcowell/ev3dev-lang-python/blob/4bc425f90270bd1fbdfbbc2daaf8c1e88728056d/ev3dev/motor.py#L729).

**Pull request merged**